File: sdk/typescript/src/values.ts
// Conversion between JavaScript values and DataGit's canonical wire values.
//
// This file exists because of one hazard, and everything in it follows from
// that hazard: JavaScript's `number` is a float64. It cannot hold every int64,
// and it cannot hold an exact decimal at all. A value sent through DataGit is
// HASHED INTO HISTORY, so a value that changes shape on the way through changes
// the commit id for data the user never edited.
//
// So the mapping is explicit and refuses the ambiguous cases:
//
//   bigint   -> int          exact 64-bit integers
//   Decimal  -> numeric      exact decimals, carried as a string
//   number   -> float        ONLY where a float is what the column holds
//
// A `number` for a decimal column is refused rather than silently rounded.

import type { Value } from "./gen/datagit/v1/datagit_pb.js";
import { timestampFromDate, timestampDate } from "@bufbuild/protobuf/wkt";

/** An exact decimal. A string, because that is the only JavaScript type that
 *  holds one without loss. */
export class Decimal {
  constructor(readonly text: string) {
    if (!/^-?\d+(\.\d+)?([eE][-+]?\d+)?$/.test(text)) {
      throw new TypeError(`not a decimal: ${JSON.stringify(text)}`);
    }
  }
  toString(): string {
    return this.text;
  }
}

/** dec is shorthand for building a Decimal. */
export function dec(text: string | number): Decimal {
  return new Decimal(typeof text === "number" ? String(text) : text);
}

export type Cell = null | boolean | bigint | number | Decimal | string | Uint8Array | Date;

/** toWire converts a JavaScript value to a wire Value. */
export function toWire(v: Cell): Value {
  if (v === null || v === undefined) {
    return { $typeName: "datagit.v1.Value", kind: { case: "isNull", value: true } } as Value;
  }
  if (typeof v === "boolean") {
    return { $typeName: "datagit.v1.Value", kind: { case: "boolValue", value: v } } as Value;
  }
  if (typeof v === "bigint") {
    return { $typeName: "datagit.v1.Value", kind: { case: "intValue", value: v } } as Value;
  }
  if (v instanceof Decimal) {
    return {
      $typeName: "datagit.v1.Value",
      kind: { case: "numericValue", value: v.text },
    } as Value;
  }
  if (typeof v === "number") {
    if (!Number.isFinite(v)) {
      throw new TypeError(
        `cannot send ${v}: NaN and Infinity have no canonical encoding, because ` +
          `NaN is not equal to itself and a hash needs equality to mean something`,
      );
    }
    return { $typeName: "datagit.v1.Value", kind: { case: "floatValue", value: v } } as Value;
  }
  if (typeof v === "string") {
    return { $typeName: "datagit.v1.Value", kind: { case: "textValue", value: v } } as Value;
  }
  if (v instanceof Uint8Array) {
    return { $typeName: "datagit.v1.Value", kind: { case: "bytesValue", value: v } } as Value;
  }
  if (v instanceof Date) {
    return {
      $typeName: "datagit.v1.Value",
      kind: { case: "timeValue", value: timestampFromDate(v) },
    } as Value;
  }
  throw new TypeError(
    `cannot send ${typeof v} to DataGit. Use bigint for exact integers, ` +
      `dec("1.25") for exact decimals, Date for timestamps, or Uint8Array for binary`,
  );
}

/** fromWire converts a wire Value to a JavaScript value. */
export function fromWire(v: Value): Cell {
  switch (v.kind?.case) {
    case "isNull":
    case undefined:
      return null;
    case "boolValue":
      return v.kind.value;
    case "intValue":
      // A bigint, not a number: an int64 past 2^53 loses precision as a number,
      // and DataGit's sequence numbers reach into that range.
      return v.kind.value;
    case "floatValue":
      return v.kind.value;
    case "numericValue":
      return new Decimal(v.kind.value);
    case "textValue":
      return v.kind.value;
    case "bytesValue":
      return v.kind.value;
    case "timeValue":
      return timestampDate(v.kind.value);
  }
  throw new Error(`unknown value kind ${(v.kind as { case?: string })?.case}`);
}

File: sdk/python/datagit/values.py
"""Conversion between Python values and DataGit's canonical wire values.

The mapping is EXPLICIT and total. There is no "figure out what this is" path,
because the one ambiguous case -- a number -- is the case that matters most:
sending a Python float where the column holds an exact decimal is how a value
silently changes on its way into a hash.
"""

from __future__ import annotations

import datetime as _dt
from decimal import Decimal
from typing import Any

from google.protobuf.timestamp_pb2 import Timestamp

from .v1 import datagit_pb2 as pb


def to_wire(v: Any) -> pb.Value:
    """Convert a Python value to a wire Value.

    `Decimal` becomes an exact numeric carried as a string. `float` becomes a
    float and is REFUSED for anything that needs exactness -- if the column is a
    decimal, pass a Decimal.
    """
    if v is None:
        return pb.Value(is_null=True)
    if isinstance(v, bool):  # before int: bool is an int in Python
        return pb.Value(bool_value=v)
    if isinstance(v, int):
        return pb.Value(int_value=v)
    if isinstance(v, Decimal):
        return pb.Value(numeric_value=str(v))
    if isinstance(v, float):
        return pb.Value(float_value=v)
    if isinstance(v, str):
        return pb.Value(text_value=v)
    if isinstance(v, (bytes, bytearray)):
        return pb.Value(bytes_value=bytes(v))
    if isinstance(v, _dt.datetime):
        ts = Timestamp()
        ts.FromDatetime(v)
        return pb.Value(time_value=ts)
    raise TypeError(
        f"cannot send {type(v).__name__} to DataGit. Use Decimal for exact "
        f"numbers, datetime for timestamps, or bytes for binary"
    )


def from_wire(v: pb.Value) -> Any:
    """Convert a wire Value to a Python value."""
    which = v.WhichOneof("kind")
    if which is None or which == "is_null":
        return None
    if which == "bool_value":
        return v.bool_value
    if which == "int_value":
        return v.int_value
    if which == "float_value":
        return v.float_value
    if which == "numeric_value":
        # A Decimal, not a float. Round-tripping through float would lose the
        # exactness the wire format went to trouble to preserve.
        return Decimal(v.numeric_value)
    if which == "text_value":
        return v.text_value
    if which == "bytes_value":
        return v.bytes_value
    if which == "time_value":
        return v.time_value.ToDatetime()
    raise ValueError(f"unknown value kind {which}")
